File: 2020/22_CrabCombat/test_player.py
# ======================================================================
# Crab Combat
#   Advent of Code 2020 Day 22 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                    t e s t _ p l a y e r . p y
# ======================================================================
"Test solver for Advent of Code 2020 day 22, Crab Combat"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import unittest

import player

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
EXAMPLE = {'number': 1, 'cards': [9, 2, 6, 3, 1]}

# ======================================================================
#                                                             TestPlayer
# ======================================================================


class TestPlayer(unittest.TestCase):  # pylint: disable=R0904
    "Test Player object"

    def test_empty_init(self):
        "Test the default Player creation"

        # 1. Create default Player object
        myobj = player.Player()

        # 2. Make sure it has the default values
        self.assertEqual(myobj.part2, False)
        self.assertEqual(myobj.number, 0)
        self.assertEqual(len(myobj.cards), 0)

    def test_text_init(self):
        "Test the Player object creation from text"

        # 1. Create Player object from text
        myobj = player.Player(number=EXAMPLE['number'], cards=EXAMPLE['cards'])

        # 2. Make sure it has the expected values
        self.assertEqual(myobj.part2, False)
        self.assertEqual(myobj.number, 1)
        self.assertEqual(len(myobj.cards), 5)

        # 3. Test methods
        self.assertEqual(myobj.score(), 45 + 8 + 18 + 6 + 1)
        self.assertEqual(myobj.get_top_card(), 9)
        self.assertEqual(len(myobj.cards), 4)
        myobj.keep(9, 5)
        self.assertEqual(len(myobj.cards), 6)
        self.assertEqual(myobj.score(), 12 + 30 + 12 + 3 + 18 + 5)
        self.assertEqual(myobj.lost(), False)
        self.assertEqual(myobj.get_top_card(), 2)
        self.assertEqual(myobj.get_top_card(), 6)
        self.assertEqual(myobj.get_top_card(), 3)
        self.assertEqual(myobj.get_top_card(), 1)
        self.assertEqual(myobj.get_top_card(), 9)
        self.assertEqual(myobj.get_top_card(), 5)
        self.assertEqual(myobj.get_top_card(), None)
        self.assertEqual(myobj.get_top_card(), None)
        self.assertEqual(myobj.lost(), True)


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                     t e s t _ p l a y e r . py                 end
# ======================================================================

File: 2020/22_CrabCombat/test_player.lua
-- ======================================================================
-- Crab Combat
--   Advent of Code 2020 Day 22 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                    t e s t _ p l a y e r . l u a
-- ======================================================================
-- Test Player for Advent of Code 2020 day 22, Crab Combat

-- ----------------------------------------------------------------------
--                                                                require
-- ----------------------------------------------------------------------
local luaunit = require('luaunit')

local player = require('player')

-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------

-- ======================================================================
--                                                             TestPlayer
-- ======================================================================

function test_empty_init()
  -- Test the default Player creation

  -- 1. Create default Player object
  local myobj = player:Player()

  -- 2. Make sure it has the default values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(myobj.number, 0)
  luaunit.assertEquals(#myobj.cards, 0)
  luaunit.assertEquals(tostring(myobj), '0: ')

end

function test_text_init()
  -- Test the Player object creation from text

  -- 1. Create Game object from text
  local myobj = player:Player({number=1})

  -- 2. Make sure it has the expected values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(myobj.number, 1)
  luaunit.assertEquals(#myobj.cards, 0)
  
  -- 3. Test methods
  myobj:add_cards({9, 2, 6, 3, 1})
  luaunit.assertEquals(#myobj.cards, 5)
  luaunit.assertEquals(tostring(myobj), '1: 9, 2, 6, 3, 1')
  luaunit.assertEquals(myobj:score(), 45 + 8 + 18 + 6 + 1)
  luaunit.assertEquals(myobj:get_top_card(), 9)
  luaunit.assertEquals(#myobj.cards, 4)
  myobj:keep(9, 5)
  luaunit.assertEquals(#myobj.cards, 6)
  luaunit.assertEquals(myobj:score(), 12 + 30 + 12 + 3 + 18 + 5)
  luaunit.assertEquals(myobj:is_deck_empty(), false)  
  luaunit.assertEquals(myobj:get_top_card(), 2)
  luaunit.assertEquals(myobj:get_top_card(), 6)
  luaunit.assertEquals(myobj:get_top_card(), 3)
  luaunit.assertEquals(myobj:get_top_card(), 1)
  luaunit.assertEquals(myobj:get_top_card(), 9)
  luaunit.assertEquals(myobj:get_top_card(), 5)
  luaunit.assertEquals(myobj:is_deck_empty(), true)  
end

function test_clone()
  -- Test the cloning a player

  -- 1. Create Game object from text
  local myobj = player:Player({number=1})

  -- 2. Make sure it has the expected values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(myobj.number, 1)
  luaunit.assertEquals(#myobj.cards, 0)
  
  -- 3. Test methods
  myobj:add_cards({9, 2, 6, 3, 1})
  luaunit.assertEquals(#myobj.cards, 5)
  luaunit.assertEquals(tostring(myobj), '1: 9, 2, 6, 3, 1')
  
  local newobj = myobj:clone(3)
  luaunit.assertEquals(newobj.part2, false)
  luaunit.assertEquals(newobj.number, 1)
  luaunit.assertEquals(#newobj.cards, 3)
  luaunit.assertEquals(tostring(newobj), '1: 9, 2, 6')
end

  
-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------
os.exit( luaunit.LuaUnit.run() )

-- ======================================================================
-- end                 t e s t _ p l a y e r . p y                end
-- ======================================================================
